timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/database/tinkerpop/Neo4jIndexHandler.java

```diff
@@ -66,31 +87,48 @@ public void addToOrUpdateQuickSearchIndex(Collection collection, String quickSea
   public void removeFromQuickSearchIndex(Collection collection, Vertex vertex) {
     Index<Node> index = getFulltextIndex(collection.getCollectionName());
 
-    index.remove(graphDatabase().getNodeById((long) vertex.id()), QUICK_SEARCH);
+    // make sure only this field is removed from the index.
+    index.remove(vertexToNode(vertex), QUICK_SEARCH);
   }
 
-  private GraphTraversal<Vertex, Vertex> traversalFromIndex(Collection collection, QuickSearch quickSearch) {
-    Index<Node> index = getFulltextIndex(getIndexName(collection));
-    IndexHits<Node> hits = index.query(QUICK_SEARCH, createQuery(quickSearch));
+  @Override
+  public GraphTraversal<Vertex, Vertex> findById(UUID timId) {
+    IndexHits<Node> hits = getExactIndex(ID_INDEX).query(TIM_ID, timId.toString());
     List<Long> ids = StreamSupport.stream(hits.spliterator(), false).map(h -> h.getId()).collect(toList());
 
     return ids.isEmpty() ? EmptyGraphTraversal.instance() : traversal().V(ids);
   }
 
-  private GraphTraversalSource traversal() {
-    return tinkerpopGraphManager.getGraph().traversal();
+  @Override
+  public void addToIdIndex(UUID timId, Vertex vertex) {
+    Index<Node> index = getExactIndex(ID_INDEX);
+    index.add(vertexToNode(vertex), TIM_ID, timId.toString());
   }
 
-  private Index<Node> getFulltextIndex(String collectionName) {
-    // Add the config below, to make sure the index is case insensitive.
-    Map<String, String> indexConfig = MapUtil.stringMap(IndexManager.PROVIDER, "lucene", "type", "fulltext");
-    return indexManager().forNodes(collectionName, indexConfig);
+
+  @Override
+  public void addToOrUpdateIdIndex(UUID timId, Vertex vertex) {
```

Comment:
Maybe the best names are: `insertInto(id|Quicksearch)Index` and `upsertInto(Id|Quicksearch)index` then the extra documentation is no longer needed I guess. Even if you expect the index to handle only one vertex per lookup value InsertInto would warn you that you need to check out what this method does.